api/database/dataRun.js

```diff
@@ -0,0 +1,212 @@
+            
+            // Get a schema of module-key pairs from data_run i
+            const oldSchema = await this.getSchema(i, DeviceData);
+
+            // If there is no schema for data_run i, check the next
+            if (oldSchema === null || oldSchema === undefined) {
+                continue;
+            }
+            else {
+                // If any oldSchema exists, this is not the first entry
+                console.log("Found an oldSchema for data run:", i);
+                firstEntry = false;
+            }
+
+            // First check if the schemas have the same number of sensors
+            if (newSchema.length !== oldSchema.length) {
+                console.log("New NumElements: ", newSchema.length);
+                console.log("Old NumElements: ", oldSchema.length);
+                console.log ("Num sensors does not match schema for data_run ", i);
+                continue;
+            }
+            else {
+                const sameSchema = await this.compareModules(newSchema, oldSchema);
+                // If the new schema matches one prior, set the dataRun and return it
+                if (sameSchema === true) {
+                    console.log("Schemas match at dataRun ", i);
+                    this.dataRun = i;
+                    return this;
+                }
+                // Otherwise, check the next oldSchema
+            }
+        }
+
+        // If this is the first entry, don't change the dataRun
+        if (firstEntry) {
+            console.log("This is the first data entry for this device.");
+            this.num_dataRuns = 1;
+            return this;
+        }
+        else {
+            // newSchema is different from all past schemas, so assign unique dataRun
+            this.num_dataRuns = this.num_dataRuns + 1;
+            this.dataRun = this.num_dataRuns;
+            return this;
+        }
+    }
+
+    /**
+	 * Compares the schemas of two separate deviceData objects.
+	 *
+	 * 1. Fills an array of strings for each module-key pair
+     * 2. Compares the arrays for equality.
+	 *
+	 * @param {Array} newSchema - The data contents array of the deviceData
+     * object which contains modules and key-value pairs for data points.
+	 * @param {Array} oldSchema - The data contents array of an old deviceData
+     * object which contains modules and key-value pairs for data points.
+	 * @returns {{boolean}} - True if they have the same schemas (every
+     * module-key pair is identical)
+	 */
+    async compareModules (newSchema, oldSchema) {
+
+        let j;
+        // Check every sensor in the arrays
+        for (j = 0; j < newSchema.length; j++) {
+
+            let newArray = await this.modulekey_Array(j, newSchema);
+            let oldArray = await this.modulekey_Array(j, oldSchema);
+
+            // If schemas don't have same number of keys, they are not identical
+            if (newArray.length !== oldArray.length) {
```

Comment:
It looks like you're comparing by length twice. Here and at line 69. Do you need to check it both times?

Reply:
Yes. The first is comparing the number of modules in the contents array. This is comparing the number of module-key pairs for each module j.